### PR TITLE
linux: enable EFI_ZBOOT for generic compression support

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

Starting 6.1, it is possible to have generic compression for aarch64 and riscv64 via CONFIG_EFI_ZBOOT which implements apparently its own decompression.

https://lwn.net/Articles/907679/

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 1]
```diff
@@ -891,7 +891,8 @@ let
 
       EFI_STUB            = yes; # EFI bootloader in the bzImage itself
       EFI_GENERIC_STUB_INITRD_CMDLINE_LOADER =
-          whenOlder "6.2" (whenAtLeast "5.8" yes); # initrd kernel parameter for EFI
+        whenOlder "6.2" (whenAtLeast "5.8" yes); # initrd kernel parameter for EFI
```
Comment (RaitoBezarius) [inline pkgs/os-specific/linux/kernel/common-config.nix:894]: will get rid of this change

**Comment by RaitoBezarius** [comment]
I am testing this by cross compiling a RISCV64 kernel first.

**Comment by alyssais** [comment]
> aarch64 and riscv64

And loongarch!

**Comment by alyssais** [comment]
Needs unixtools.hexdump in `nativeBuildInputs`.

**Comment by RaitoBezarius** [comment]
@alyssais Were you able to measure the reduced size? I don't know if I'm holding this wrong but I cannot see anything else than `Image` being produced for aarch64 or riscv64 with the same size.

**Comment by arianvp** [comment]
I think we also need to change the `target` here to `bzImage` (or `zImage` ?):

https://github.com/NixOS/nixpkgs/blob/df708463c90db17e8e5984b4ff98094f791622eb/lib/systems/platforms.nix#L379

**Comment by RaitoBezarius** [comment]
> I think we also need to change the `target` here to `bzImage` (or `zImage` ?):
> 
> https://github.com/NixOS/nixpkgs/blob/df708463c90db17e8e5984b4ff98094f791622eb/lib/systems/platforms.nix#L379

Yeah, actually, I am planning for a bigger refactor of that.

**Comment by ius** [comment]
> @alyssais Were you able to measure the reduced size? I don't know if I'm holding this wrong but I cannot see anything else than `Image` being produced for aarch64 or riscv64 with the same size.

For reference, some numbers for my Aarch64 cross compiled kernel:
- `Image`: 56M
- `vmlinuz.efi` 15M

This would be really useful for undersized `/boot` partitions.

**Comment by zzywysm** [comment]
Commenting because this is important and shouldn't be marked stale.

**Comment by flokli** [comment]
I'm also interested in this, as `rust-hypervisor` is not happy with the too large kernels we currently produce. I summarized the situation in https://b.tvl.fyi/issues/407.

**Comment by flokli** [comment]
> I'm also interested in this, as `rust-hypervisor` is not happy with the too large kernels we currently produce. I summarized the situation in https://b.tvl.fyi/issues/407.

I was just dumb, and used the wrong kernel output (from the `dev` output). `${myKernel}/${stdenv.hostPlatform.linux-kernel.target}` actually has a somewhat reasonable size on aarch64-linux too. This is probably still important to get the number down further?

**Comment by fpletz** [comment]
I've been successfully booting my Pinebook Pro with Tow-Boot, systemd-boot and zboot. :rocket: These were the changed needed if somebody wants to test on another aarch64 machine: https://github.com/NixOS/nixpkgs/compare/master...fpletz:nixpkgs:arm64-zboot

To get this PR through the the door, I have fixed the formatting issue, rebased and added a commit restricting EFI_ZBOOT to aarch64 for now since I have tested the kernel builds correctly both the Image and the vmlinuz.efi. This is needed since otherwise the x86 build breaks. We should add new systems when they have tested and thus not inadvertently break kernel builds of non-release architectures.

Current file sizes for aarch64:

```
70M	/boot/EFI/nixos/jvs0n8skfvzxixbaw74qmjz3df6m0znf-linux-6.9.7-Image.efi
17M	/boot/EFI/nixos/qgx88i268aq4ra1sd6xjzhdmsmxybny3-linux-6.9.7-vmlinuz.efi.efi
```

The extra `.efi` is added by `systemd-boot-builder.py`. :smirk:

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 2]
```diff
@@ -1051,6 +1051,8 @@ let
       EFI_STUB            = yes; # EFI bootloader in the bzImage itself
       EFI_GENERIC_STUB_INITRD_CMDLINE_LOADER =
           whenOlder "6.2" (whenAtLeast "5.8" yes); # initrd kernel parameter for EFI
+       # Generic compression support for EFI payloads
```
Comment (flokli) [inline pkgs/os-specific/linux/kernel/common-config.nix:1054]: Can you update this comment to explain why we restrict it to aarch64 for now? Just so someone stumbling over just the file (not the discussion in the PR) understands why it's there.

Comment (fpletz) [reply pkgs/os-specific/linux/kernel/common-config.nix]: Added a comment. Thanks!

**Comment by flokli** [comment]
Thanks for carrying this over the finish line! 🎉

**Comment by arianvp** [comment]
But this PR as is doesn't do anything as we still `make install` the _uncompressed_ kernel I think? We need to make sure we do `make vmlinuz.efi` as opposed to `make Image` on aarch64

https://github.com/NixOS/nixpkgs/pull/287122 fixes that iirc

**Comment by arianvp** [comment]
Ah https://github.com/NixOS/nixpkgs/pull/324419 is the follow-up. I missed that

**Pull request merged**